### PR TITLE
[`flake8-pytest-style`] Add automatic fix for `pytest-parametrize-values-wrong-type` (`PT007`)

### PR DESCRIPTION
**Pull request by autinerd**
## Summary

This adds automatic fixes for the `PT007` rule.

I am currently reviewing and adding Ruff rules to Home Assistant. One rule is PT007, which has multiple hundred occurrences in the codebase, but no automatic fix, and this is not fun to do manually, especially because using Regexes are not really possible with this.

My knowledge of the Ruff codebase and Rust in general is not good and this is my first PR here, so I hope it is not too bad.

One thing where I need help is: How can I have the transformed code to be formatted automatically, instead of it being minimized as it does it now?

## Test Plan

Using the existing fixtures and updated snapshots.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Thanks for this!

> One thing where I need help is: How can I have the transformed code to be formatted automatically, instead of it being minimized as it does it now?

So, if you want to do this, you can't use `generator`. The generator is just a basic printer that takes an AST and prints it without formatting. If you want to preserve the existing formatting (including comments), you need to implement a more manual fix (e.g., change the `[` to `(` and the `]` to `)`, or similar). Check out `unnecessary_generator_set` for an example of what that might look like.

I'm fine to merge with the generator-based fix, though a locator-based fix tends to preserve more of this kind of trivia. Are you interested in trying it?

**Comment by autinerd** [comment]
Oh yes, I see it now that this approach deletes all comments. Thanks for pointing me to a place to check, I will try if I can do this.

**Comment by charliermarsh** [comment]
Yeah it's kind of a tradeoff. The generator is much easier (to program, and to get right), but much lower-fidelity.

**Comment by autinerd** [comment]
So, I have it now as a locator-based fix based of `unnecessary_generator_set`, thanks!

It also checks if a list has only 1 element and transforms `["test"]` into `("test",)` to make it a valid tuple.

The only thing is, that `("test",)` will be converted into `["test",]`, which is valid from the syntax, but of course not the most beautiful.

**Comment by autinerd** [comment]
And I have it now as a unsafe fix, but I think it should be safe to use in this approach.

**Comment by autinerd** [comment]
Oh, ruff-format seems to transform `["test",]` unconditionally into
```
[
    "test",
]
```

Then maybe I need to remove the comma, because this disrupts the source formatting.

**Comment by charliermarsh** [comment]
You could, e.g., _only_ remove the trailing comma if it immediately precedes the closing brace.

**Comment by autinerd** [comment]
I'm sorry, but can you maybe hint me on how to access the text of the `Expr` to that I can check on whether it ends with `,]`?

**Comment by autinerd** [comment]
It doesn't look very elegant in my opinion, but it works.

**Review by charliermarsh — APPROVED** [review]

File: crates/ruff_linter/resources/test/fixtures/flake8_pytest_style/PT007.py [hunk 1]
```diff
@@ -79,5 +79,6 @@ def test_single_list_of_lists(param):
 
 @pytest.mark.parametrize("a", [1, 2])
 @pytest.mark.parametrize(("b", "c"), ((3, 4), (5, 6)))
+@pytest.mark.parametrize("d", [3,])
```
Comment (charliermarsh) [inline crates/ruff_linter/resources/test/fixtures/flake8_pytest_style/PT007.py:82]: Just one bugfix: the PR was adding a trailing comma here, which led to `,,`

**Pull request merged**